irods_client/query_operations.py

```diff
@@ -0,0 +1,218 @@
+    # - distinct (optional): Set to 1 to collapse duplicate rows, otherwise set to 0. Defaults to 1. Only supported by GenQuery 1
+    # - parser (optional): User either genquery1 or genquery2. Defaults to genquery1.
+    # - sql_only (optional): Set to 1 to execute an SQL only query, otherwise set to 0. Defaults to 0. Only supported by GenQuery2.
+    # - zone (optional): The zone name. Defaults ot the local zone.
+    # return
+    # - Status code 2XX: The results of the GenQuery.
+    # - Other: Status code and return message.
+    def execute_genquery(self, query: str, offset: int=0, count: int=-1, case_sensitive: int=1, distinct: int=1,
+                            parser: str='genquery1', sql_only: int=0, zone: str=''):
+        if (self.token == None):
+            raise Exception('No token set. Use setToken() to set the auth token to be used')
+        if (not isinstance(query, str)):
+            raise Exception('query must be a string')
+        if ((not isinstance(offset, int)) or (not offset >= 0)):
+            raise Exception('offset must be an int greater than 0')
+        if ((not isinstance(count, int)) or (not count >= -1)):
+            raise Exception('count must be an int greater than 0 or flag value -1')
+        if ((not case_sensitive == 0) and (not case_sensitive == 1)):
+            raise Exception('case_sensitive must be an int 1 or 0')
+        if ((not distinct == 0) and (not distinct == 1)):
+            raise Exception('distinct must be an int 1 or 0')
+        if ((not parser == 'genquery1') and (not parser == 'genquery2')):
+            raise Exception('parser must be either \'genquery1\' or \'genquery2\'')
+        if ((not sql_only == 0) and (not sql_only == 1)):
+            raise Exception('sql_only must be an int 1 or 0')
+        if (not isinstance(zone, str)):
+            raise Exception('zone must be a string')
+
+        headers = {
+            'Authorization': 'Bearer ' + self.token,
+        }
+
+        params = {
+            'op': 'execute_genquery',
+            'query': query,
+            'offset': offset,
+            'parser': parser
+        }
+
+        if (count != -1):
+            params['count'] = count
+
+        if (zone != ''):
+            params['zone'] = zone
+        
+        if (parser == 'genquery1'):
+            params['case-sensitive'] = case_sensitive
+            params['distinct'] = distinct
+        else:
+            params['sql-only'] = sql_only
+
+        r = requests.get(self.url_base + '/query', headers=headers, params=params)
+
+        if (r.status_code / 100 == 2):
+            rdict = r.json()
+
+            if rdict['irods_response']['status_code']:
+                print('Failed to execute query: iRODS Status Code' + str(rdict['irods_response']['status_code']))
+            else:
+                print('Query executed successfully')
+            
+            return(rdict)
+        else:
+            print('Error: ' + r.text)
+
+            return(r)
```

Comment:
With your current design, users cannot react to the HTTP status code or iRODS status code.

Users need access to that information. Please re-examine and ask questions if it's unclear how to proceed.

Reply:
Did some testing and ran into an issue. I set it up this way to prevent errors from trying to call .json() when the response wasn't a 2xx. The content is always empty for these requests from what I can tell. I'll keep trying some other things out to see though.

Reply:
consider writing the use cases first (as tests)... and then... the right answer may present itself.

Reply:
I did find a case where the content was not empty for a 400 response. I'm thinking of having the wrapper return the content anyways even if it's empty so that the user can always have it, possibly in a dict along with the status code.

Reply:
That sounds better.

Consider speaking with the other interns to see how they are approaching this.

Reply:
I took a look at the other wrappers and decided to go with this idea. It's pretty much the same as how it's handled in the typescript wrapper.

Reply:
Looking at the latest commit, it appears that your logic is not balanced. In the success case (HTTP status code of 2XX), the HTTP status code is not included. In the failure case, it is included. That's going to make it difficult for users if they cannot assume the base properties aren't included in the value returned by the function call.

Reply:
The wrapper now always returns a response in the form:

```
{
    'status_code': r.status_code,
    'data': r.json() or r.text depending on the status returned
}
```

Unit tests have been updated to accommodate this change and are passing.

Reply:
Okay. Mark this conversation as resolved if you feel it has been addressed.